### PR TITLE
Reorganize ray_deps_setup.bzl to make all the GitHub rules uniform and download ZIP files for everything

### PR DESCRIPTION
**Pull request by mehrdadn**
## Why are these changes needed?

- ZIP file downloads are much faster than git clones, due to the lack of history, resulting in much faster build times
- The build rules have redundancy and are becoming disorganized and hard to maintain

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

**Review by jovany-wang — COMMENTED** [review]
This PR looks a good enhancement, but is it really necessary to define `github_repository` to do the conversion?

What do you think of declaring them by `http_archive` simply?  I think it's  clearer than `github_repository()`.

Thanks!

**Comment by mehrdadn** [comment]
@jovany-wang Yes, it's necessary. `http_archive` has been a source of redundancy and disorganization, and previously there has been a need for `git_repository` as well.

Compare what it was like before the change to what it's like after. Before the change, we had:

    RULES_JVM_EXTERNAL_TAG = "2.10"
    RULES_JVM_EXTERNAL_SHA = "1bbf2e48d07686707dd85357e9a94da775e1dbd7c464272b3664283c9c716d26"
    http_archive(
        name = "rules_jvm_external",
        sha256 = RULES_JVM_EXTERNAL_SHA,
        strip_prefix = "rules_jvm_external-%s" % RULES_JVM_EXTERNAL_TAG,
        url = "https://github.com/bazelbuild/rules_jvm_external/archive/%s.zip" % RULES_JVM_EXTERNAL_TAG,
    )

Now we have:

    github_repository(
        name = "rules_jvm_external",
        tag = "2.10",
        remote = "https://github.com/bazelbuild/rules_jvm_external",
        sha256 = "1bbf2e48d07686707dd85357e9a94da775e1dbd7c464272b3664283c9c716d26",
    )

**Comment by jovany-wang** [comment]
@mehrdadn  I see.  It simplified the process of catting remote address and version-tag to url.

**Comment by mehrdadn** [comment]
@jovany-wang Yeah-- and also simplified the process of switching between git and zip and local repos, and also removed the redundancy in `build_file` and `strip_prefix`, and also made it easier to test locally with `path =` (fewer merge conflicts)... it simplified a fair number of things I think. :-)

File: bazel/ray_deps_setup.bzl [hunk 1]
```diff
@@ -1,128 +1,198 @@
 load("@bazel_tools//tools/build_defs/repo:git.bzl", "git_repository", "new_git_repository")
 load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
 
-def ray_deps_setup():
-    RULES_JVM_EXTERNAL_TAG = "2.10"
-    RULES_JVM_EXTERNAL_SHA = "1bbf2e48d07686707dd85357e9a94da775e1dbd7c464272b3664283c9c716d26"
+def github_repository(**kwargs):
+    """
+    Conveniently chooses between ZIP, git, etc. GitHub repositories.
+    We prefer ZIP downloads, since they're smaller and faster
```
Comment (jovany-wang) [inline bazel/ray_deps_setup.bzl]:
```suggestion
    Recommend to download zip packages, because zip packages have no history of the repositories.
```

File: bazel/ray_deps_setup.bzl [hunk 2]
```diff
@@ -1,128 +1,198 @@
 load("@bazel_tools//tools/build_defs/repo:git.bzl", "git_repository", "new_git_repository")
 load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
 
-def ray_deps_setup():
-    RULES_JVM_EXTERNAL_TAG = "2.10"
-    RULES_JVM_EXTERNAL_SHA = "1bbf2e48d07686707dd85357e9a94da775e1dbd7c464272b3664283c9c716d26"
+def github_repository(**kwargs):
```
Comment (jovany-wang) [inline bazel/ray_deps_setup.bzl]: How about naming it `ray_dependency_repository()`, because we use this method not only declaring github dependencies, but also declaring other type dependencies.

File: bazel/ray_deps_setup.bzl (hunk 1 above)
Comment (jovany-wang) [reply bazel/ray_deps_setup.bzl]: BTW, the comments of this method seems too verbose, could you refine them?

File: bazel/ray_deps_setup.bzl (hunk 2 above)
Comment (mehrdadn) [reply bazel/ray_deps_setup.bzl]: I'm not sure what you mean -- what other type dependencies is this used for? Note that this is intended strictly for GitHub, and it's not specific to Ray in any way -- you can't in general switch between ZIP archives and git clones; it's only possible on GitHub because they have a well-defined URL format.

File: bazel/ray_deps_setup.bzl (hunk 1 above)
Comment (mehrdadn) [reply bazel/ray_deps_setup.bzl]: I agree it's more verbose for my liking too, but I've in fact already rewritten it several times to get it into this shape. :) I removed the parameters that are already described in Bazel and only documented any additional knowledge needed. I'm not sure what else to condense without removing pertinent information to be honest. I'm happy to consider an alternative if you have an suggestions on how to shorten it, though.

File: bazel/ray_deps_setup.bzl [hunk 3]
```diff
@@ -1,128 +1,198 @@
...
+            kwargs.setdefault("strip_prefix", "%s-%s" % (project, treeish))
+        if not sha256:
+            # git clone
+            if commit != None: kwargs.setdefault("commit", commit)
+            if tag != None: kwargs.setdefault("tag", tag)
+            if branch != None: kwargs.setdefault("branch", branch)
+            kwargs.setdefault("shallow_since", shallow_since)
+            kwargs.setdefault("remote", remote)
+            if "build_file" in kwargs:
+                new_git_repository(**kwargs)
```
Comment (edoakes) [inline bazel/ray_deps_setup.bzl]: please just call this and `git_repository` below with expanded keyword arguments for clarity, otherwise I have to trace back all of the code that modifies `kwargs` to understand what this line is passing in

Comment (edoakes) [reply bazel/ray_deps_setup.bzl]: and `http_archive` below

File: bazel/ray_deps_setup.bzl [hunk 4]
```diff
@@ -1,128 +1,198 @@
...
+    If sha256 is absent  , obtains clone via git (slow and NOT recommended)
+
+    strip_prefix is automatic by default.
+    Set strip_prefix=<string> (or False) to override.
+
+    Some common parameters (like shallow_since)
+    are automatically stripped when not applicable.
+    """
+
+    remote = kwargs.pop("remote")  # required
```
Comment (edoakes) [inline bazel/ray_deps_setup.bzl]: if this is required it shouldn't be a kwarg

**Review by edoakes — CHANGES_REQUESTED** [review]
I'm not a huge fan of having magic default parameters because it is hard to debug when things go wrong (you basically have to reverse-engineer this function). Would be better to make things explicit wherever possible.

Also it would be great to remove all occurrences of the kwargs dictionary in this function. Better to just define the required arguments and optional keyword arguments and then pass them into the other functions we're calling instead of manually modifying kwargs in place, it's very hard to follow.

File: bazel/ray_deps_setup.bzl [hunk 5]
```diff
@@ -1,128 +1,198 @@
...
+
+    The remote parameter is required.
+    One of {commit, tag, branch} is also required.
+    The name parameter is optional but recommended.
+    Set sha256 = True if the ZIP archive hash is unknown, then update it ASAP
+    to allow caching & avoid repeated downloads on every build.
+
+    If build_file is a string, it refers to a custom build file.
+    If build_file is absent, the project is assumed to contain a build file.
+    If build_file == True, then the build file name is auto-deduced.
```
[the remark this answers is not in the document]
Comment (mehrdadn) [reply bazel/ray_deps_setup.bzl]: @edoakes Yeah, it's a lot of redundancy for no gain. It doesn't affect reproducibility in any way.

File: bazel/ray_deps_setup.bzl [hunk 6]
```diff
@@ -1,128 +1,198 @@
...
 load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
 
-def ray_deps_setup():
-    RULES_JVM_EXTERNAL_TAG = "2.10"
-    RULES_JVM_EXTERNAL_SHA = "1bbf2e48d07686707dd85357e9a94da775e1dbd7c464272b3664283c9c716d26"
+def github_repository(**kwargs):
+    """
+    Conveniently chooses between ZIP, git, etc. GitHub repositories.
+    We prefer ZIP downloads, since they're smaller and faster
+    due to the lack of history.
```
[the remark this answers is not in the document]
Comment (mehrdadn) [reply bazel/ray_deps_setup.bzl]: We're not calling git manually here, Bazel is calling it. And it performs a lot worse than archive downloads.

File: bazel/ray_deps_setup.bzl [hunk 7]
```diff
@@ -1,128 +1,198 @@
...
+    Set sha256 = True if the ZIP archive hash is unknown, then update it ASAP
+    to allow caching & avoid repeated downloads on every build.
+
+    If build_file is a string, it refers to a custom build file.
+    If build_file is absent, the project is assumed to contain a build file.
+    If build_file == True, then the build file name is auto-deduced.
+
+    If path   is present , local repository is assumed.
+    If sha256 == <string>, obtains ZIP download with the given hash (fast)
+    If sha256 == True    , obtains ZIP download without knowing the hash (fast)
```
[the remark this answers is not in the document]
Comment (mehrdadn) [reply bazel/ray_deps_setup.bzl]: Yes, as follows:

    Set sha256 = True if the ZIP archive hash is unknown, then update it ASAP
    to allow caching & avoid repeated downloads on every build.

File: bazel/ray_deps_setup.bzl (hunk 4 above)
Comment (mehrdadn) [reply bazel/ray_deps_setup.bzl]: We don't want it to be positional. That's not normal Bazel practice for dependency rules. It would also interfere with the canonical ordering of arguments.

File: bazel/ray_deps_setup.bzl [hunk 8]
```diff
@@ -1,128 +1,198 @@
...
+    strip_prefix is automatic by default.
+    Set strip_prefix=<string> (or False) to override.
+
+    Some common parameters (like shallow_since)
+    are automatically stripped when not applicable.
+    """
+
+    remote = kwargs.pop("remote")  # required
+
+    GIT_SUFFIX = ".git"  # We need to strip this
```
[the remark this answers is not in the document]
Comment (mehrdadn) [reply bazel/ray_deps_setup.bzl]: That's wrong though. `"foobartig.git".rstrip(".git") == "foobar"`

File: bazel/ray_deps_setup.bzl (hunk 2 above)
[the remark this answers is not in the document]
Comment (mehrdadn) [reply bazel/ray_deps_setup.bzl]: I can if you really want to, but are do you really prefer having to update this function every time someone wants to use an argument that wasn't previously in the supported list? I tried to make this transparent so it would just pass unknown arguments through.

File: bazel/ray_deps_setup.bzl (hunk 3 above)
Comment (mehrdadn) [reply bazel/ray_deps_setup.bzl]: Hmm okay, if you confirm https://github.com/ray-project/ray/pull/6193/files#r348190605 then I'll do this.

**Review by mehrdadn — COMMENTED** [review]
@edoakes 

Edit: Not sure why GitHub is showing two copies of my comments, sorry if that's confusing. :\

File: bazel/ray_deps_setup.bzl (hunk 5 above)
[the remark this answers is not in the document]
Comment (mehrdadn) [reply bazel/ray_deps_setup.bzl]: Yeah, it's a lot of redundancy for no gain. It doesn't affect reproducibility in any way.

File: bazel/ray_deps_setup.bzl [hunk 9]
```diff
@@ -1,128 +1,198 @@
...
+    One of {commit, tag, branch} is also required.
+    The name parameter is optional but recommended.
+    Set sha256 = True if the ZIP archive hash is unknown, then update it ASAP
+    to allow caching & avoid repeated downloads on every build.
+
+    If build_file is a string, it refers to a custom build file.
+    If build_file is absent, the project is assumed to contain a build file.
+    If build_file == True, then the build file name is auto-deduced.
+
+    If path   is present , local repository is assumed.
```
[the remark this answers is not in the document]
Comment (mehrdadn) [reply bazel/ray_deps_setup.bzl]: Yes, I use it for development (as can anyone who may want to try patching a dependency). It's for local copies of repositories, so it of course doesn't make sense to be used for checked-in code.

File: bazel/ray_deps_setup.bzl (hunk 2 above)
[the remark this answers is not in the document]
Comment (edoakes) [reply bazel/ray_deps_setup.bzl]: Yes that would be preferred

File: bazel/ray_deps_setup.bzl (hunk 6 above)
[the remark this answers is not in the document]
Comment (edoakes) [reply bazel/ray_deps_setup.bzl]: Looks like bazel calls with depth=1 if `tag` is specified:
https://docs.bazel.build/versions/0.25.0/repo/git.html

File: bazel/ray_deps_setup.bzl (hunk 8 above)
[the remark this answers is not in the document]
Comment (edoakes) [reply bazel/ray_deps_setup.bzl]: wow, that is an insane design choice

File: bazel/ray_deps_setup.bzl (hunk 4 above)
Comment (edoakes) [reply bazel/ray_deps_setup.bzl]: In that case can we just check that it's set and raise a useful error?

File: bazel/ray_deps_setup.bzl (hunk 7 above)
[the remark this answers is not in the document]
Comment (edoakes) [reply bazel/ray_deps_setup.bzl]: Is it used by any of our build targets though? No point in adding it if not and this seems like an easy way to accidentally end up with slow builds.

File: bazel/ray_deps_setup.bzl (hunk 2 above)
[the remark this answers is not in the document]
Comment (mehrdadn) [reply bazel/ray_deps_setup.bzl]: Ok I'll change it

File: bazel/ray_deps_setup.bzl (hunk 4 above)
Comment (mehrdadn) [reply bazel/ray_deps_setup.bzl]: Ok

File: bazel/ray_deps_setup.bzl (hunk 6 above)
[the remark this answers is not in the document]
Comment (mehrdadn) [reply bazel/ray_deps_setup.bzl]: Tags don't have any integrity checking though; clones do.
Also, I saw maybe 2 tags out of the numerous dependencies we have; everything else was by commit hash. I don't think you can realistically assume every commit we want from every dependency will necessarily have a tag for us to use.

File: bazel/ray_deps_setup.bzl (hunk 8 above)
[the remark this answers is not in the document]
Comment (mehrdadn) [reply bazel/ray_deps_setup.bzl]: Yeah. I think the reason they allow it to be a string is for efficiency, but I think they should've forced it to be a keyword argument if the caller wanted to provide more than one character.

File: bazel/ray_deps_setup.bzl (hunk 7 above)
[the remark this answers is not in the document]
Comment (mehrdadn) [reply bazel/ray_deps_setup.bzl]: No what I mean is that it should never be checked in (similar to how `local` will be checked in), but it is used locally to download the ZIP without knowing the hash a priori. And as a bonus Bazel tells you what the hash is when you do that, so it's quite easy to just copy-paste from the terminal that's already in front of you. Otherwise you have force everyone (including yourself) to manually download the ZIP in the browser and run `sha256sum` before you can add a dependency at all. And then you have to have Bazel download it one _more_ time now that it has the hash. Which is both quite painful in itself, and also a time-waster given that some ZIPs are fairly large and now have to be downloaded twice.

File: bazel/ray_deps_setup.bzl (hunk 2 above)
[the remark this answers is not in the document]
Comment (mehrdadn) [reply bazel/ray_deps_setup.bzl]: @edoakes Wait I just realized I can't do that, because if I put them here then people will try to call them positionally rather than as keywords. And Starlark (which is this Python-like language) forbids keyword-only arguments, so I can't do that either.

That said though, I'll try to shuffle the logic to make it easier to see what the defaults are.

[the remark this answers is not in the document]
Comment (mehrdadn) [reply bazel/ray_deps_setup.bzl]: Update: see new commits.

File: bazel/ray_deps_setup.bzl (hunk 3 above)
Comment (mehrdadn) [reply bazel/ray_deps_setup.bzl]: Update: I think this isn't workable anymore in light of https://github.com/ray-project/ray/pull/6193#discussion_r348218943.

File: bazel/ray_deps_setup.bzl [hunk 10]
```diff
@@ -1,131 +1,194 @@
...
+    """
+    Conveniently chooses between archive, git, etc. GitHub repositories.
+    Prefer archives, as they're smaller and faster due to the lack of history.
+
+    One of {commit, tag, branch} must also be provided (as usual).
+
+    sha256 should be omitted (or None) when the archive hash is unknown, then
+    updated ASAP to allow caching & avoid repeated downloads on every build.
+
+    NOTE: False and None mean different things.
```
Comment (edoakes) [inline bazel/ray_deps_setup.bzl]: don't need this, it's a python convention :)

File: bazel/ray_deps_setup.bzl [hunk 11]
```diff
@@ -1,131 +1,194 @@
...
+    If sha256       != False, uses archive download (recommended; fast).
+    If sha256       == False, uses git clone (NOT recommended; slow).
+    If path         != None , local repository is assumed at the given path.
+    """
+    GIT_SUFFIX = ".git"
+    archive_suffix = ".zip"
+
+    treeish = commit or tag or branch
+    if not treeish: fail("Missing commit, tag, or branch argument")
+    if remote == None: fail("Missing remote argument")
```
Comment (edoakes) [inline bazel/ray_deps_setup.bzl]: `if remote is None`

File: bazel/ray_deps_setup.bzl [hunk 12]
```diff
@@ -1,131 +1,194 @@
...
+
+    project = (remote[:-len(GIT_SUFFIX)]  # We need to strip this suffix
+               if remote.endswith(GIT_SUFFIX)
+               else remote).split("//", 1)[1].split("/")[2]
```
Comment (edoakes) [inline bazel/ray_deps_setup.bzl]: This is pretty hard to read, would be nice if you could you separate out the logic a bit into a few lines and add some comments about how exactly we're mangling this string

**Review by edoakes — APPROVED** [review]
Awesome, this looks great!

File: bazel/ray_deps_setup.bzl (hunk 11 above)
Comment (mehrdadn) [reply bazel/ray_deps_setup.bzl]: @edoakes I wish. Starlark requires `==`. :(

File: bazel/ray_deps_setup.bzl (hunk 12 above)
Comment (mehrdadn) [reply bazel/ray_deps_setup.bzl]: Done

File: bazel/ray_deps_setup.bzl (hunk 10 above)
Comment (mehrdadn) [reply bazel/ray_deps_setup.bzl]: Alrighty :P
